### PR TITLE
exec: add sumInt64Agg

### PR DESCRIPTION
**Pull request by asubiotto**
This operator consumes a column of bools and a column of int64s, where the bools indicate a new group. For each group, this operator accumulates the sum of the associated int64s.

This operator will be templated later.

```
BenchmarkSumInt/groupSize=1/numInputBatches=1-8         	  200000	      6643 ns/op	1233.13 MB/s
BenchmarkSumInt/groupSize=1/numInputBatches=2-8         	  100000	     12689 ns/op	1291.11 MB/s
BenchmarkSumInt/groupSize=1/numInputBatches=32-8        	   10000	    190247 ns/op	1377.91 MB/s
BenchmarkSumInt/groupSize=1/numInputBatches=64-8        	    5000	    401202 ns/op	1306.79 MB/s
BenchmarkSumInt/groupSize=2/numInputBatches=1-8         	  200000	      6610 ns/op	1239.29 MB/s
BenchmarkSumInt/groupSize=2/numInputBatches=2-8         	  100000	     12240 ns/op	1338.46 MB/s
BenchmarkSumInt/groupSize=2/numInputBatches=32-8        	   10000	    177775 ns/op	1474.58 MB/s
BenchmarkSumInt/groupSize=2/numInputBatches=64-8        	    5000	    351800 ns/op	1490.30 MB/s
BenchmarkSumInt/groupSize=512/numInputBatches=1-8       	  200000	      6737 ns/op	1215.90 MB/s
BenchmarkSumInt/groupSize=512/numInputBatches=2-8       	  200000	     11735 ns/op	1396.10 MB/s
BenchmarkSumInt/groupSize=512/numInputBatches=32-8      	   10000	    161915 ns/op	1619.02 MB/s
BenchmarkSumInt/groupSize=512/numInputBatches=64-8      	    5000	    321052 ns/op	1633.03 MB/s
BenchmarkSumInt/groupSize=1024/numInputBatches=1-8      	  200000	      6707 ns/op	1221.38 MB/s
BenchmarkSumInt/groupSize=1024/numInputBatches=2-8      	  200000	     11795 ns/op	1388.99 MB/s
BenchmarkSumInt/groupSize=1024/numInputBatches=32-8     	   10000	    159267 ns/op	1645.93 MB/s
BenchmarkSumInt/groupSize=1024/numInputBatches=64-8     	    5000	    318549 ns/op	1645.86 MB/s
```

### CODE REVIEW

**Comment by asubiotto** [comment]
First two commits are #31908 and #31906. The `FromColumnTypes` commit is just extracting some code that I can see us reusing quite often.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31924)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/sum_int64_agg.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6QZeEQOw59ao1xuF:-LPl6QZf1iGVwVU4NQZc:b-h9xq58) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L79)):*
> ```Go
> 	curIdx := -1
> 	// Re-initialize the output batch.
> 	for i := uint16(0); i < a.outputBatchSize; i++ {
> ```

Wondering if there's a way to get rid of this.
___
*[pkg/sql/exec/sum_int64_agg.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6Shz3va4XpaipEuo:-LPl6Shz3va4XpaipEup:b-u6hfak) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L109)):*
> ```Go
> 			// TODO(asubiotto): Modifying the input batch is probably not cool.
> 			if groups[firstIdx] {
> 				groups[firstIdx] = false
> ```

This might need to be changed somehow. Open to suggestions.
___
*[pkg/sql/exec/sum_int64_agg.go, line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6YCbBIR8_sVfWKwa:-LPl6YCbBIR8_sVfWKwb:b4uf2iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L129)):*
> ```Go
> 				a.scratch.vec[curIdx] += toSum
> 				x := 0
> 				if groups[i] {
> ```

Add a comment about the if statement being optimized out
___
*[pkg/sql/exec/sum_int64_agg.go, line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6_bb0wnLo7_B63HJ:-LPl6_bb0wnLo7_B63HK:bk9jwkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L137)):*
> ```Go
> 		} else {
> 			i := uint16(resumeIdx)
> 			for ; i < batch.Length() && curIdx < int(a.outputBatchSize); i++ {
> ```

Is it cool to have two conditions on these loops?
___
*[pkg/sql/exec/sum_int_test.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6m0UCyKqEjXSZLRL:-LPl6m0UCyKqEjXSZLRM:b-aysm3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int_test.go#L26)):*
> ```Go
> 
> func TestSumInt(t *testing.T) {
> 	testCases := []struct {
> ```

More tests?
___
*[pkg/sql/exec/utils_test.go, line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6qqY80Kcmog_t7UC:-LPl6qqY80Kcmog_t7UD:bwehcfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/utils_test.go#L133)):*
> ```Go
> 
> func (s *opTestInput) Next() ColBatch {
> 	batch := NewMemBatch(append(s.typs, s.extraCols...))
> ```

This is the wrong commit but serves the same purpose. Make sure to get rid of it before merging (once expected behavior gets resolved)

**Review by jordanlewis — CHANGES_REQUESTED** [review]
Cool stuff!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-:-LPlK84i9qxBXcSuSIU0:b95tn33)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/sum_int64_agg.go, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlJJ1ZCTImOtl7WYD4:-LPlJJ1ZCTImOtl7WYD5:b-mq3ipj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L56)):*
> ```Go
> }
> 
> func (a *sumInt64Agg) Init() {
> ```

This should also initialize the operator's parent. I know it's not done that way everywhere in the code right now, but I'm switching things to act like that (it's a lot simpler).
___
*[pkg/sql/exec/sum_int64_agg.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6QZeEQOw59ao1xuF:-LPlEwMGBMXk7IK4Kp0p:bu8f06r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L79)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Wondering if there's a way to get rid of this.
</blockquote></details>

You should `copy` a zero integer vector into your scratch vector which might be faster. The same thing is done in the distinct operator. I originally had a whole operator for this, `zero`, but I think it's easier to just have individual operators do the copying themselves. Define a global var that's always zeroed to copy from.
___
*[pkg/sql/exec/sum_int64_agg.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlHgAB9uEtdtW5On5h:-LPlHgAB9uEtdtW5On5i:b-ghoi9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L90)):*
> ```Go
> 		}
> 
> 		groups, ints, sel := batch.ColVec(0).Bool(), batch.ColVec(1).Int64(), batch.Selection()
> ```

So the original idea we had (all temp columns for an "operator group" (e.g all the sub aggregators for an aggregator) belong inside of the colbatch) is actually more confusing then it's worth. I think you were right to question it.

I think it's just a waste of bookkeeping to try to maintain that, especially when you can just share a concrete vector between all of the operators in an "operator group" for the scratch column.

I fixed that for distinct in https://github.com/cockroachdb/cockroach/pull/31881. Please take a look at that and see what you think, I'm hoping the pattern makes sense and that we can use it here. I do think it greatly simplifies the code.
___
*[pkg/sql/exec/sum_int64_agg.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6Shz3va4XpaipEuo:-LPlIY-L2A-KbEI0FjG2:b-65bw2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L109)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This might need to be changed somehow. Open to suggestions.
</blockquote></details>

See my suggestion above - `groups` should be a raw `bool` column that's shared between all the aggregators in a group. And sorry if that was your original idea and I discouraged it - my thinking has changed :)
___
*[pkg/sql/exec/sum_int64_agg.go, line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6_bb0wnLo7_B63HJ:-LPlJilq90sy7NcMKyNn:b-r0gl2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L137)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is it cool to have two conditions on these loops?
</blockquote></details>

There's ways to improve this, performance wise, but I'd focus on getting a full first cut first. The main approach to take is slicing the vectors you're looking at precisely. So you'd say `groups = groups[resumeIdx:batch.Length()]` and iterate over that with `range`.
___
*[pkg/sql/exec/sum_int_test.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlIyEHDW27NRq9TIxc:-LPlIyEHDW27NRq9TIxd:bppi4ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int_test.go#L180)):*
> ```Go
> 				func(b *testing.B) {
> 					// Only count the int64 column.
> 					// Because an a.Next() call returns
> ```

might have cut off this sentence
___
*[pkg/sql/exec/types/types.go, line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlJ3WSDL7sdS3LMThK:-LPlJ3WSDL7sdS3LMThL:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/types/types.go#L80)):*
> ```Go
> 
> // FromColumnTypes calls FromColumnType on each element of cts, returning the
> // resulting slice.
> ```

:thumbsup:

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31924)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/sum_int64_agg.go, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlJJ1ZCTImOtl7WYD4:-LPlx-V_3gkGiBRMiQfM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L56)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This should also initialize the operator's parent. I know it's not done that way everywhere in the code right now, but I'm switching things to act like that (it's a lot simpler).
</blockquote></details>

Done.
___
*[pkg/sql/exec/sum_int64_agg.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6QZeEQOw59ao1xuF:-LPmCa8GC2FKnS84AROL:b-oea35j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L79)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You should `copy` a zero integer vector into your scratch vector which might be faster. The same thing is done in the distinct operator. I originally had a whole operator for this, `zero`, but I think it's easier to just have individual operators do the copying themselves. Define a global var that's always zeroed to copy from.
</blockquote></details>

Nice:
```
name                                        old time/op    new time/op    delta
SumInt/groupSize=1/numInputBatches=1-8        6.98µs ± 9%    5.67µs ± 9%  -18.81%  (p=0.000 n=9+9)
SumInt/groupSize=1/numInputBatches=2-8        12.5µs ± 2%    11.1µs ± 5%  -11.18%  (p=0.000 n=9+9)
SumInt/groupSize=1/numInputBatches=32-8        188µs ± 1%     177µs ±14%     ~     (p=0.139 n=8+9)
SumInt/groupSize=1/numInputBatches=64-8        378µs ± 5%     349µs ±10%   -7.67%  (p=0.013 n=9+10)
SumInt/groupSize=2/numInputBatches=1-8        6.64µs ± 3%    5.47µs ± 3%  -17.61%  (p=0.000 n=9+9)
SumInt/groupSize=2/numInputBatches=2-8        11.7µs ± 4%    10.4µs ± 1%  -11.19%  (p=0.000 n=9+8)
SumInt/groupSize=2/numInputBatches=32-8        177µs ± 2%     163µs ± 2%   -7.63%  (p=0.000 n=9+9)
SumInt/groupSize=2/numInputBatches=64-8        349µs ±10%     324µs ± 2%   -7.05%  (p=0.002 n=10+10)
SumInt/groupSize=512/numInputBatches=1-8      6.66µs ± 2%    5.35µs ± 3%  -19.68%  (p=0.000 n=10+10)
SumInt/groupSize=512/numInputBatches=2-8      11.6µs ± 2%    10.3µs ± 3%  -11.39%  (p=0.000 n=10+10)
SumInt/groupSize=512/numInputBatches=32-8      157µs ± 2%     159µs ± 5%     ~     (p=0.050 n=9+9)
SumInt/groupSize=512/numInputBatches=64-8      325µs ± 7%     324µs ± 8%     ~     (p=0.905 n=9+10)
SumInt/groupSize=1024/numInputBatches=1-8     6.81µs ±11%    5.36µs ± 2%  -21.20%  (p=0.000 n=10+10)
SumInt/groupSize=1024/numInputBatches=2-8     11.4µs ± 2%    10.3µs ± 1%  -10.21%  (p=0.000 n=10+10)
SumInt/groupSize=1024/numInputBatches=32-8     157µs ± 1%     156µs ± 2%   -0.95%  (p=0.022 n=9+10)
SumInt/groupSize=1024/numInputBatches=64-8     315µs ± 3%     314µs ± 2%     ~     (p=0.278 n=9+10)

name                                        old speed      new speed      delta
SumInt/groupSize=1/numInputBatches=1-8      1.17GB/s ± 9%  1.45GB/s ± 8%  +23.17%  (p=0.000 n=9+9)
SumInt/groupSize=1/numInputBatches=2-8      1.31GB/s ± 2%  1.46GB/s ±14%  +11.02%  (p=0.002 n=9+10)
SumInt/groupSize=1/numInputBatches=32-8     1.40GB/s ± 1%  1.49GB/s ±12%     ~     (p=0.139 n=8+9)
SumInt/groupSize=1/numInputBatches=64-8     1.39GB/s ± 5%  1.51GB/s ±10%   +8.57%  (p=0.013 n=9+10)
SumInt/groupSize=2/numInputBatches=1-8      1.23GB/s ± 3%  1.50GB/s ± 3%  +21.40%  (p=0.000 n=9+9)
SumInt/groupSize=2/numInputBatches=2-8      1.40GB/s ± 4%  1.57GB/s ± 4%  +11.95%  (p=0.000 n=9+9)
SumInt/groupSize=2/numInputBatches=32-8     1.48GB/s ± 2%  1.60GB/s ± 2%   +8.27%  (p=0.000 n=9+9)
SumInt/groupSize=2/numInputBatches=64-8     1.51GB/s ± 9%  1.62GB/s ± 2%   +7.36%  (p=0.002 n=10+10)
SumInt/groupSize=512/numInputBatches=1-8    1.23GB/s ± 2%  1.53GB/s ± 3%  +24.51%  (p=0.000 n=10+10)
SumInt/groupSize=512/numInputBatches=2-8    1.42GB/s ± 2%  1.60GB/s ± 2%  +12.83%  (p=0.000 n=10+10)
SumInt/groupSize=512/numInputBatches=32-8   1.67GB/s ± 2%  1.64GB/s ± 6%   -2.31%  (p=0.028 n=9+10)
SumInt/groupSize=512/numInputBatches=64-8   1.62GB/s ± 7%  1.62GB/s ± 7%     ~     (p=0.905 n=9+10)
SumInt/groupSize=1024/numInputBatches=1-8   1.21GB/s ±10%  1.53GB/s ± 2%  +26.59%  (p=0.000 n=10+10)
SumInt/groupSize=1024/numInputBatches=2-8   1.43GB/s ± 2%  1.60GB/s ± 1%  +11.37%  (p=0.000 n=10+10)
SumInt/groupSize=1024/numInputBatches=32-8  1.67GB/s ± 1%  1.68GB/s ± 2%   +0.97%  (p=0.022 n=9+10)
SumInt/groupSize=1024/numInputBatches=64-8  1.66GB/s ± 3%  1.67GB/s ± 2%     ~     (p=0.278 n=9+10)
```
___
*[pkg/sql/exec/sum_int64_agg.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlHgAB9uEtdtW5On5h:-LPmDbkK8AaViiqVcXh7:b-tmykna) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L90)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

So the original idea we had (all temp columns for an "operator group" (e.g all the sub aggregators for an aggregator) belong inside of the colbatch) is actually more confusing then it's worth. I think you were right to question it.

I think it's just a waste of bookkeeping to try to maintain that, especially when you can just share a concrete vector between all of the operators in an "operator group" for the scratch column.

I fixed that for distinct in https://github.com/cockroachdb/cockroach/pull/31881. Please take a look at that and see what you think, I'm hoping the pattern makes sense and that we can use it here. I do think it greatly simplifies the code.
</blockquote></details>

Pattern makes sense. Talked offline about this for a bit and will integrate the same sort of approach when I write the parent aggregator (aka big daddy).
___
*[pkg/sql/exec/sum_int64_agg.go, line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6_bb0wnLo7_B63HJ:-LPmP_GkB8coKfUCg8-X:bkd5ega) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int64_agg.go#L137)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

There's ways to improve this, performance wise, but I'd focus on getting a full first cut first. The main approach to take is slicing the vectors you're looking at precisely. So you'd say `groups = groups[resumeIdx:batch.Length()]` and iterate over that with `range`.
</blockquote></details>

Added a TODO
___
*[pkg/sql/exec/sum_int_test.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6m0UCyKqEjXSZLRL:-LPmQC4S3tO0A3wYC4oB:b-9860ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int_test.go#L26)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

More tests?
</blockquote></details>

I think I've covered all the important cases.
___
*[pkg/sql/exec/sum_int_test.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPlIyEHDW27NRq9TIxc:-LPmRLqSCR4-VNX64C2O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/sum_int_test.go#L180)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

might have cut off this sentence
</blockquote></details>

Done.
___
*[pkg/sql/exec/utils_test.go, line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-LPl6qqY80Kcmog_t7UC:-LPmRJq-BppkLsg0MagY:bb7abui) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dcc011347618fd73cc20edcbb18a554f9a9a62c/pkg/sql/exec/utils_test.go#L133)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is the wrong commit but serves the same purpose. Make sure to get rid of it before merging (once expected behavior gets resolved)
</blockquote></details>

Got rid of commit.

**Comment by asubiotto** [comment]
ping.

I believe I have your LGTM but want to make sure.

**Review by jordanlewis — APPROVED** [review]
:lgtm: sorry about that!

> Reviewed 1 of 5 files at r5, 2 of 2 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31924#-:-LQ0Uw7KBphN1VDiFSTg:b-e18f63)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by asubiotto** [comment]
bors r=jordanlewis

**Pull request merged**
